### PR TITLE
fix(cassandra) ensure single coordinator in migrations

### PR DESCRIPTION
**Pull request by thibaultcha**
### Summary

This ensures we respect the proper 'single coordinator' pattern for
migrations when one of them is using the DAO's `find_all()` method.

### Full changelog

* ensure `find_all()` only uses the migrations coordinator.
* ensure we wait for schema consensus before doing so, since the
  coordinator will have to get responses about the table's content from
  its peers.
* perf: only wait for schema consensus if we ran some migrations at all.
* properly pass C* timeout values set in the Kong configuration to the
driver.
* new `cassandra_schema_consensus_timeout` property.
* bump lua-cassandra to 1.2.1 which ensures we update the Nginx time
when testing for a schema consensus timeout.

**NOTE**: Do **NOT** merge yet.

### CODE REVIEW

File: kong/dao/db/cassandra.lua [hunk 1]
```diff
@@ -389,7 +390,22 @@ function _M:find_all(table_name, tbl, schema)
...
 
-  for rows, page_err in self.cluster:iterate(query, args, opts) do
+  local iter = self.cluster.iterate
+  local iter_self = self.cluster
+  if coordinator then
+    -- we are in migrations, and need to wait for a schema consensus
+    -- before performing such a DML query
+    local ok, err = self:wait_for_schema_consensus()
+    if not ok then
+      return nil, "could not wait for schema consensus: " .. err
```
Comment (Tieske) [inline kong/dao/db/cassandra.lua]: "failed waiting for schema consensus"

File: kong/dao/factory.lua [hunk 2]
```diff
@@ -299,13 +299,15 @@ function _M:run_migrations(on_migrate, on_success)
...
-      return ret_error_string(self.db.name, nil,
-                              "could not wait for schema consensus: " .. err)
+    if migrations_ran > 0 then
+      log.verbose("now waiting for schema consensus (%dms) timeout",
+                  self.db.cluster.max_schema_consensus_wait)
+
+      local ok, err = self.db:wait_for_schema_consensus()
+      if not ok then
+        return ret_error_string(self.db.name, nil,
+                                "could not wait for schema consensus: " .. err)
```
Comment (Tieske) [inline kong/dao/factory.lua]: "failed waiting for..."

Comment (thibaultcha) [inline kong/dao/factory.lua]: both forms are used inter-dependently in the codebase

**Comment by thibaultcha** [comment]
The "do not merge" label is for *refused* PRs.

> Would it be possible to add a test for the behaviour?

Sadly not, or it would be included.

**Comment by thibaultcha** [comment]
> as the magic seems to be in the driver.

Actually the driver has little to do with this all.

**Comment by Tieske** [comment]
> The "do not merge" label is for refused PRs.

surely we close those? this label at least is more descriptive than a foot note in the original post.

File: kong/dao/factory.lua (hunk 2 above)
Comment (Tieske) [inline kong/dao/factory.lua]: I'd say that only impatient people 'cannot wait' 😄

**Comment by thibaultcha** [comment]
> surely we close those?

It is sometimes more complicated than that.

> I'd say that only impatient people 'cannot wait' 😄

Sounds just like the definition of a timeout!

**Comment by thibaultcha** [comment]
> considering updating the error messages optional.

Updated them in the end :wink:

**Pull request merged**